### DISCUSSION TITLE
Making MT32 capable whdload titles work...

### DISCUSSION THREAD

**Discussion by giantclambake** [BlitterStudio/amiberry#1361] [General]
As MT32 emulation is a latterly addition to amiberry, I really didn't expect the XML to cater for this case.

AFAICT, there are <20 whdload.lha game packs so disposed (I didn't check itch.io), so I'm unsure if there's much impetus to change the status-quo.

One can relatively easily create a config.uae file to be associated with these whdload.lha titles, to enable MT32 playback, which makes me wonder if this is more fodder for `wiki/FAQ` section....(sometimes I think the wiki needs a 'mini how-to' section for stuff like this =)

OTOH.... if we really wanted the whd-booter process to automate this..... snippet example;

```
       <game filename="CastleOfDrBrain_v1.1_De_MT32_0304" sha1="4799ab40eb4888878af002e27b58df39d98c1415">
                <name>Castle Of Dr Brain De MT 32</name>
                <subpath>CastleOfDrBrainDeMT32</subpath>
                <variant_uuid>15c5fca7-2d83-509c-a76d-0c77f5952f3b</variant_uuid>
                <slave_count>1</slave_count>
                <slave_default>CastleOfDrBrainDeMT32.Slave</slave_default>
                <slave_libraries>False</slave_libraries>
                <slave number="1">
                        <filename>CastleOfDrBrainDeMT32.Slave</filename>
                        <datapath>data</datapath>
                </slave>
                <hardware>
                        PRIMARY_CONTROL=JOYSTICK
                        PORT0=JOY
                        PORT1=JOY
                        SCREEN_AUTOHEIGHT=TRUE
                        SCREEN_CENTERH=SMART
                        SCREEN_CENTERV=SMART
                </hardware>
        </game>
```

...we would need an additional `<hardware>` field of something like `SOUND=MIDI` ~ amiberry's task would be to enable MT32 emulation (playback only) if this field is present.

What drove my thoughts this way, is there is a very clear differentiation wrt the .lha game packs themselves... they are all available in both MT32 & non-MT32 versions, identified by the 'MT32' hint ....ie;

`CastleOfDrBrain_v1.1_De_0304.lha`   as opposed to  `CastleOfDrBrain_v1.1_De_MT32_0304.lha`

To me, this gives rise to the premise that an amiberry user, who full well knows they have the MT32 roms present & MT32 emulation works, may deliberately choose the MT32 versions of these whdload.lha titles, fully expecting the whdload/whd-booter processes to enable MT32 emulation automatically....yes? (that is, I would NOT expect them to have to create a config.uae for the title, to get MT32 sound to work =)

Thoughts?

TIA

**Comment by midwan** [comment]
Either we add something like that to the XML, and have Amiberry pick it up, or we try to detect it from the filename and auto-magically enable it from Amiberry's side when relevant.

We do something similar with CD32 titles, so assuming the filenames contain it, it wouldn't be too hard to do the same for MT32.

**Comment by giantclambake** [comment]
I thought about that, but in essence it relies on the user not changing the filename (and possibly removing the 'MT32' hint).

The XML process, by comparison, identifies the whdload.lha by it's MD5sum, thus making it filename agnostic.... that's why it was my first choice ; the situation is already 'well targeted' as such.

As I say though, I could only find 17 whdload archives so disposed (and HOL only lists 22 Amiga titles in all that support MT32), so it's a very small group of titles...I honestly don't think it's worthwhile reworking the XML for such a small number of titles.

Still think it's a nice little feature add tho' ...up to you. We could do it by filename detection if that's easy, and if any user ever asks "Why doesn't MT32 work for my whdload title?", you know the first response is "Did you change the filename?" .... ;)

**Reply by midwan** [reply]
@HoraceAndTheSpider 
Since you're more up-to-date with the XML management, what's your input on this subject? How much work would it be to add a new entry in the hardware there, for those titles? I'm guessing there won't be any new ones coming in the future...

**Reply by midwan** [reply]
@giantclambake 
I've added file-based detection for now, not sure if we'll change the XML for these few titles in the future. If we do, it's an easy-enough change to pick it up from there as well.

**Reply by giantclambake** [reply]
@midwan 
Sounds like a plan to me ~ I'll give it a quick test before I crawl under the covers...

**Comment by giantclambake** [comment]
Seems to work as expected, uncovered oddity wrt MT32 emulation...

![ex](https://imgur.com/436iLk8.png)

![ex2](https://imgur.com/alcNCQN.png)

Sound's great however, and works ... so that's a win, thanks ;)

**Reply by midwan** [reply]
Yeah, the sysex messages will output to a real MIDI device (if connected), otherwise to the console as shown above. :)

**Reply by giantclambake** [reply]
Understood ~ if you want to see the worst of that wrt console, try  `./amiberry CastleOfDrBrain_v1.1_De_MT32_0304.lha`

**Comment by giantclambake** [comment]
Closing this as it's implemented ;)
